[WIP] Websockets

vincentjames501 opened the pull request:
This isn't intended to be merged, but to start a discussion about what a potential websocket API could look like using JDK11's HttpClient.

gnarroway commented:
Thanks for putting this together. I think as a first cut I am slightly reluctant to add the manifold dependency and impose this library on consumers. I would prefer to provide the minimal websocket building blocks, and the manifold (or core.async) stuff can initially be documented as usage examples or provided as a separate library, and pulled in eventually as usage patterns become more evident. That said it does look quite nice to use. 

- `websocket` takes a single map of uri, listener, http-client, connect-timeout-ms etc and returns a WebSocket. 
- `http-client` defaults to `(HttpClient/newHttpClient)`
- `send!` dispatches on data type (you’ve put this inside `websocket`)
- omit the ‘send’ prefix on everything. 
- remove the state getter fns (isInputClosed etc). May replace with some map representation of the ws or datafy at some point. 

Happy to keep this PR lying around for documentation purposes so the above can go in a separate one. Otherwise I can do it in the new year (I can’t push to github from my work laptop I am currently traveling with)

vincentjames501 commented:
I agree with the reluctance as I felt the same way working with it. I do think a different package under the hato namespace would be nice for something like that. 

> send! dispatches on data type (you’ve put this inside websocket)

I played around with protocols and multimethods but different event types like ping, pong, etc felt "weird" to represent with it so it may be best to leave those as separate functions (especially as they have different argument types like `last?`)? If the only real things to dispatch on is binary and text, it may not be worth the abstraction?

Manifold complicated things as you can only put certain things into the stream so sending things like ping can make it messy which is why I had two variants of the websocket manifold implementation where one supported low level events.

I split the manifold stuff into a separate namespace for now to make it simpler to look over for now, used datafy instead, etc.

Feel free to psuedocode out something that dispatch on data type for sending. Curious what you're thinking exactly

> omit the ‘send’ prefix on everything.

I changed `send-ping!` and `send-pong!` to `ping!`/`pong!` as those are verbs but left it as `send-text!`/`send-binary!` as those feel weird to just call `binary!` for example.

> remove the state getter fns (isInputClosed etc). May replace with some map representation of the ws or datafy at some point.

Done

gnarroway commented:
Great. Re merging the sends, I meant only the text and binary, the same as you already had: 
```clojure
(cond
        (instance? CharSequence msg)
        (websocket/send-text! ws msg true)....
```

‘(defn send! ([ws data]) ([ws data {:keys [last?] :or {last? true}}]))’
